Added Wikipedia's own Dark Mode solution

THEBOSS0369 opened the pull request:
This PR Fixes #1304

This PR implements wikipedia's dark mode css styling into kiwix app. Fixed the alignment of the search bar as well it is now in the center. Below are the ss for reference

I have tested it on 4 - 5 zims and it worked on every zim below are the ss (the image inversion has become more good which isn't in the default dark theme). 
List 
**Wikipedia astronomy** latest (as you suggested)

Before
![image](https://github.com/user-attachments/assets/313a329c-6d54-4ed0-94f9-113d39e89307)
![image](https://github.com/user-attachments/assets/3a8d4e7b-9e4f-49a1-92a8-c4572daf3505)

After
![image](https://github.com/user-attachments/assets/a10dd201-51e4-402c-86c0-4c89b63f7552)
![image](https://github.com/user-attachments/assets/bc7c34e3-302d-4138-ba02-d8b7de92450f)

**termux**
Before
![image](https://github.com/user-attachments/assets/750ba155-751d-45a0-a549-5a2b7316bc30)

After
![image](https://github.com/user-attachments/assets/ca5e6390-4d25-4b93-8ed6-594a8432e060)

**wikivoyage**
Before
![image](https://github.com/user-attachments/assets/38e8de04-0b58-4619-a312-b0411ad6cc50)

After
![image](https://github.com/user-attachments/assets/580ad458-2f4b-4355-872a-0bc8af6feb04)

bitcoin and tonedear.

## Test
I have done all the necessary test 
1. Checked in both "**Restricted**" and "**ServiceWorker**" Everything is working fine.
2. Unit tests `npm test` no issue 
6. End-to-end (e2e) `tests-e2e-firefox`-> All tests Passed.
7. extension versions with production code tested
8. Browser Test -> Microsoft Edge.

THEBOSS0369 commented:
@Jaifroid i have fixed but there 2 things that i want to clarify
1. The main dark color in the app is  dark red i wasn't able to find it , so i have override the code.
2. The main color for the content page is showing a bit of light blue i have tried to use !important as less as possible. i don't know whether this is fine or not but it only shows if the zim is not uploaded yet but when the zim is uploaded then black color shows properly, i need your suggestion let me know what you think 
this is the ss for referfence
![image](https://github.com/user-attachments/assets/e2b9e132-ca09-4d9b-bc51-31cfdf7057b7)

THEBOSS0369 commented:
I'll be fixing the codefactor tomorrow .

Jaifroid commented:
@THEBOSS0369 The dark red is just an effect to warn developers that they are viewing the app in developer mode. Just deselect "Bypass AppCache" under Troubleshooting and Development, and the app will return to normal mode so you can tes the styling without your override. However, when making any changes, to see them reflected in the development environment, you'll need to turn on Bypass AppCache again and control-refresh with DevTools open.

THEBOSS0369 commented:
> @THEBOSS0369 The dark red is just an effect to warn developers that they are viewing the app in developer mode. Just deselect "Bypass AppCache" under Troubleshooting and Development, and the app will return to normal mode so you can tes the styling without your override. However, when making any changes, to see them reflected in the development environment, you'll need to turn on Bypass AppCache again and control-refresh with DevTools open.

@Jaifroid  Ahh got it , that makes sense sir, ty for this info. Other than that everything works fine.

THEBOSS0369 commented:
@Jaifroid  sir I have removed the overriding code . Let me know if anything other modification is required.

Jaifroid commented:
Thanks! Could you update from main, and when you've pulled the changes to your local fork, be sure to run `npm install` again due to the changes made there.

Jaifroid commented:
@THEBOSS0369 Thank you for these PRs. I've tried it out, but I don't see any difference in the styling of Wikipedia articles. All I see is a change in the styling of the app. Looking at your code, I see you added style rules only to `app.css`. This file only styles the app's own UI, and most of the rules you've added there will have no effect at all because we don't use any of those rules in the app's UI.

The issue was about styling the articles in Wikipedia ZIMs. To style the articles, you would need to add a stylesheet to the incoming article when it is first loaded and before it is displayed, and this would be a separate style from the "Dark (Wikimedia invert)" style. You need to work with the styling system in the app, but as we're not just talking about inversion of the entire iframe, but about styling individual elements in Wikipedia articles, it would be a question of dynamically adding a stylesheet into the iframe's document on each article load.

Have you managed to obtain the Wikipedia dark stylesheet? As far as I can tell, the only Wikipedia theme that has a dark mode is "Vector (2022)". See screenshot:

![image](https://github.com/user-attachments/assets/fafe1057-cdeb-4c9b-81c4-c2a69713668f)

So this would need to be the name of the style in the Kiwix app's dropdown. It would make sense to have a Vector light and a Vector dark, but we could start with dark only. To add these stylesheets would likely mean removing or at least overriding some of the existing styles.

So I'm very sorry, but I don't think these PRs are on the right track currently. Let me know if you think the feature is achievable. I think it's a pretty complex one to be honest!

Jaifroid commented:
Very important consideration: make sure you're working with a recent small Wikipedia ZIM, i.e. one produced at least in March 2025. I'd suggest looking at the Wikipedia English Astronomy ZIM, but there are other smaller ones that would work just as as well for experimenting.

THEBOSS0369 commented:
Hello @Jaifroid sir i have managed to add the **wikipedia's vector css** stylesheet however its not exactly the original one from wikipedia as it has more than 5k+ lines of code and it was creating so many problems for me to handle it alone so i decided to go with the little modified one of vector. It worked properly with kiwix and has same styling. I have override some elements as their styling weren't matching like the popover and the footer section. 

I have tested it on 4 - 5 zims and it worked on every zim below are the ss (the image inversion has become more good which isn't in the default dark theme). 
List 
**Wikipedia astronomy** latest (as you suggested)

Before
![image](https://github.com/user-attachments/assets/313a329c-6d54-4ed0-94f9-113d39e89307)
![image](https://github.com/user-attachments/assets/3a8d4e7b-9e4f-49a1-92a8-c4572daf3505)

After
![image](https://github.com/user-attachments/assets/a10dd201-51e4-402c-86c0-4c89b63f7552)
![image](https://github.com/user-attachments/assets/bc7c34e3-302d-4138-ba02-d8b7de92450f)

**termux**
Before
![image](https://github.com/user-attachments/assets/750ba155-751d-45a0-a549-5a2b7316bc30)

After
![image](https://github.com/user-attachments/assets/ca5e6390-4d25-4b93-8ed6-594a8432e060)

**wikivoyage**
Before
![image](https://github.com/user-attachments/assets/38e8de04-0b58-4619-a312-b0411ad6cc50)

After
![image](https://github.com/user-attachments/assets/580ad458-2f4b-4355-872a-0bc8af6feb04)

bitcoin and tonedear.

However one zim (Tonedear) didn't worked properly i didn't quite understand why the zim isn't loading for this so i am adding the ss of that as well below. 

![image](https://github.com/user-attachments/assets/ae681f0c-4ab6-40f3-b006-897f03603a90)
![image](https://github.com/user-attachments/assets/41cbd4b8-8861-4263-9a18-f289c7e4d810)

THEBOSS0369 commented:
Also sir let me know whether now this pr is on the right track or not. If not then let me know what else is required.

Updating the pr description as well.

THEBOSS0369 commented:
Okay so i have noticed few things 

1. It itsn't able to detect the url in the zim where the wikipedia style are not loading.
```shell
Title C/localhost/www/css/kiwixJS_wikiVector.css... not found in archive.
readFile @ app.js:2641
Promise.then
handleMessageChannelMessage @ app.js:2682
navigator.serviceWorker.onmessage @ app.js:988
wombat.js:21 

GET http://localhost:5173/php.net_en_all_2024-08.zim/C/localhost/www/css/kiwixJS_wikiVector.css net::ERR_ABORTED 404 (Not Found)
``` 
 This is happening for every style . **Dark (generic invert), Dark (wikimedia invert) and Dark (Wikipedia Vector)** but except of **Dark (Wikipedia Vector)** the style inversion is working fine for others.

2. For restricted mode the conversion isn't working properly neither i was able to decode anything much from the console.

Jaifroid commented:
OK, thanks. A Wikipedia-specific dark theme doesn't need to apply to non-Wikipedia files. It's not going to work, because the Wikipedia styles are not used on random sites around the Internet. So as not to create a jarring effect, I'd suggest that Wikipedia-specific dark modes should default to standard dark mode if a non-Wikimedia file is loaded, and we add an explanation of this when the user switches styles. Let me take a look at your code to see if it's on the right tracks!

Jaifroid commented:
> ```shell
> Title C/localhost/www/css/kiwixJS_wikiVector.css... not found in archive.
> readFile @ app.js:2641
> Promise.then
> handleMessageChannelMessage @ app.js:2682
> navigator.serviceWorker.onmessage @ app.js:988
> wombat.js:21 
> 
> GET http://localhost:5173/php.net_en_all_2024-08.zim/C/localhost/www/css/kiwixJS_wikiVector.css net::ERR_ABORTED 404 (Not Found)
> ```
> 
> This is happening for every style . **Dark (generic invert), Dark (wikimedia invert) and Dark (Wikipedia Vector)** but except of **Dark (Wikipedia Vector)** the style inversion is working fine for others.

@THEBOSS0369 You're testing with a Zimit-based archive instead of a standard OpenZIM (Wikipedia) archive. Zimit archives have a utility called `wombat.js` that attempts to rewrite all links in a page to a format it can extract from the ZIM. It guesses the URL, but in this case it can't cope with the way we attach the stylesheet. Up till now, the CSS just does inversion, which works with the content of iframes too. So that error has been cosmetic. However, in your case, I believe the stylesheet needs to be added into the article itself.

But all of this is irrelevant, because we should only support Wikimedia ZIMs with the vector style, and it appears to be working there from what I can see. So you just have to implement logic to switch to a standard inversion dark mode if the user loads a non-Wikimedia ZIM. We should probably try to investigate that error in a different issue, and see if we can solve it, but it's definitely not your issue to deal with here! If you use a standard Wikipedia archive, you'll see that error isn't produced.

Jaifroid commented:
In your case, with a Wikimedia ZIM, I can see that your CSS file is correctly inserted into the article DOM. So you are indeed very much on the right tracks here. See screenshot.

![image](https://github.com/user-attachments/assets/c6029b56-8c48-4ea5-8f35-976d1fad1837)

Jaifroid commented:
With a Wikipedia ZIM, your PR seems to work in Restricted mode for me. Please double-check.

THEBOSS0369 commented:
@Jaifroid thank you sir for this detailed review. You saved me a lot of time as i was gonna spend my good amount of time to debug it and it just didn't strike in my mind that not any zim's style will work with wikipedia's style😅 

> So you just have to implement logic to switch to a standard inversion dark mode if the user loads a non-Wikimedia ZIM

 I will add this feature.

One more thing sir can you please tell me which wikipedia zim are you using or any small wiki zim which will do the work as there are many multi GB's size zim and i just can't download any multi gb size zim because it will take my device's storage plus it's gonna take a good amount of time. please let me know this will be very helpful 🙂. 

![image](https://github.com/user-attachments/assets/d9001536-ffd3-44a4-afaa-6bed5d1416ad)

Jaifroid commented:
Any of these should work (in fact if you have the space, I'd test with all three, as they have slightly different layouts for front page, and more or less complex content):

https://mirror.download.kiwix.org/zim/wikipedia/wikipedia_en_climate_change_maxi_2025-03.zim (176MB)
https://mirror.download.kiwix.org/zim/wikipedia/wikipedia_en_100_maxi_2025-03.zim (38.5MB)
https://mirror.download.kiwix.org/zim/wikipedia/wikipedia_en_ray_charles_maxi_2025-03.zim (2.9MB)

Jaifroid commented:
By the way, you can test if a wikimedia ZIM is loaded with something like:

`const wikimediaZimLoaded = /wikipedia|wikivoyage|mdwiki|wiktionary/i.test(selectedArchive.file.name);`

THEBOSS0369 commented:
@Jaifroid sir i have added the function . If a user has selected the wiki vector theme and  if the user switches to the non wiki zim then the popup will appear that this theme won't be applied in the zim and it will switch to the default dark theme. 
![image](https://github.com/user-attachments/assets/651b7888-08c2-4f74-a24f-854ee6c03b43)

Now if a non wiki zim is selected then user won't be able to select the wiki vector dark theme and it will show the message that it cannot be selected. Like this
![image](https://github.com/user-attachments/assets/89529678-a857-44a3-97f8-6bea5cc82b6f)

Also one thing i noticed sometimes the vector dark theme works in restricted and sometimes it doesn't. Can you confirm that whether this is the case for you or not? I think this might be my browser cache as when i remove cache from my browser it works fine.

Jaifroid commented:
Thanks! I get the notification when I first start the app and no ZIM is yet loaded. The option is also greyed out even though no ZIM is loaded. I think those are errors: the option should be greyed out only if `selectedArchive.file.name` is set and if it is not a Wikimedia archive. If `selectedArchive.file.name` is undefined or null, the option should still be selectable.

Jaifroid commented:
In any case, I think you should remove the notification to the user, as it doesn't add any value and is just a bit... annoying! The theme should just fall back silently to the inversion theme, and then resume working if the user subsequently loads a Wikimedia ZIM.

Jaifroid commented:
You'll probably need to add some code in the `archiveReadyCallback` function calling your theme switch handler. NB, don't include the switching code directly in `archiveReadyCallback`.

THEBOSS0369 commented:
@Jaifroid sir i have added the function `handleThemeFallback()` so that if the Vector theme is currently selected when a non Wikimedia ZIM is loaded then it automatically switches to the appropriate invert theme instead. 

Along with this the dropdown wasn't showing the appropriate theme name when we switch the zims between wiki and non wiki (**fixed it now**). Also Removed the popover functionality as you suggested . If anything else is required do let me know.

Jaifroid commented:
Thanks, and apologies for slowness, I've been away. Will test asap.

Jaifroid commented:
@THEBOSS0369 I can see that auto-switching is working, which is great 🥳. However, the UI in Config doesn't show Wikipedia Vector when switching back to a Wikimedia ZIM. It still says "Dark (generic invert)". Steps to reproduce: 1. Load a Wikimedia archive; 2. Set style to Wikipedia Vector; 3. Switch to any non-Wikimedia archive (style switches to generic invert); 4. Switch back to a Wikimedia archive (style still says generic invert, but I can see that it is in fact Wikipedia Vector).

Also, with full English Wikipedia, I'm still seeing Flash of Incorrect Style when moving between articles. It basically flashes to white. On a small Wikivoyage, I don't see the issue, but maybe because large Wikipedia is slower, we're still getting the jarring white flash.

Jaifroid commented:
Also, the asterisk appearing against the Wikipedia Vector entry doesn't seem to signify anything... So, it would be helpful to show a message to the user explaining what this tyle does and that it auto-switches, like the message displayed with "Auto (Wikimedia invert)":

![image](https://github.com/user-attachments/assets/1781ba02-6cf1-4f11-b14b-d902204d49c1)

Your message should say something like `* Only applies to Wikimedia ZIMs. Falls back to generic inversion when other archives are loaded.`

Also, please add an asterisk after "Auto (Wikimedia invert)" and "Dark (Wikimedia invert)", to indicate that these will show messages too.

THEBOSS0369 commented:
@Jaifroid Apologies for late response had to complete some commitments 😅. 

> However, the UI in Config doesn't show Wikipedia Vector when switching back to a Wikimedia ZIM

I have fixed this issue and added an asterisk as you suggested.

> Also, with full English Wikipedia, I'm still seeing Flash of Incorrect Style when moving between articles

For this issue i ain't able to fix i have tried every method but just couldn't make it work. These are few methods i tried
1. Tried to preload the dark before it loads the content 
2. Tried to override the code didn't worked out as well
3. also tried to use the code present in mwinvert.css file to prevent jarring effect
4. Added the event breakpoint to debug it more but didn't work out.

 not sure if i will be able to fix it, i will need your help here sir. Although it just appears for less than a second but still it would be better if we can remove it otherwise if u think that this won't create much effect to viewers then we can let it as it is.

THEBOSS0369 commented:
ok so I have got an idea i am trying to load the wiki dark theme when the article is loading so that it shifts to the generic dark theme and then as soon as it loads it reverts back to the vector theme.

THEBOSS0369 commented:
Okay another update

Method no
5.  i tried the theme switching it didn't worked out as well it the white screen still flashes i didn't thought earlier that switching theme will reload the page again and the white bg will show up 😓🥹. 

6.  i added the overlay to prevent the style flash  during navigation this was supposed to appear during transition and disappears as soon as content loads but it didn't work out as well

sir I think as you suggested earlier, we should create a different issue for the white flash when this pr's work is done. Let me know what you think.

THEBOSS0369 commented:
> faced this in the PWA, and the only solution I found was to hide the article document (which may be in an iframe, or may be a window/tab if the user opens an article hyperlink in a new window or tab), until the DOM is fully loaded

Hello @Jaifroid sir as u suggested i tried to fix it from this way but couldn't able to make it work mainly, i modified the **cache.js** tried to hide the content until styles load, didn't worked.

even hardcoded the html file just to see if it works or not but didn't work as expected.

Please let me know what to do further when you get some time meanwhile i will be fixing some minor ui issues i noticed in the project during this pr work.

Jaifroid commented:
OK, easiest is probably to hide the iframe contents until injection is done and stylesheet loaded. That requires the iframe itself to have a dark background colour by default, which is probably set by the background colour of the HTML page. That was the solution I adopted in the PWA, I seem to remember.

THEBOSS0369 commented:
@Jaifroid update for fois

sir thanks for your the solution u gave me but its working 50 percent means if we navigate  to different page then at first black color willl show then white flash will again appear. initially it was all white but after the changes **first black then white then original black**, for testing i am using the green color so that it can be easily identified. this is the video for reference how the latest changes are working 

https://github.com/user-attachments/assets/01f44f46-3014-482e-aa2e-9ed7b37edee8

Let me know if what you think .

Also my 4th sem is coming to end so i won't be able to work on this pr till 1 may as internal tests and practicels all are aligned this week.😅

Jaifroid commented:
> Also my 4th sem is coming to end so i won't be able to work on this pr till 1 may as internal tests and practical all are aligned this week.😅

No worries, I'm fully tied up till around 5th May anyway.

Jaifroid commented:
> also sir the codefactor test is failing (complex method) because of my change in uituitl.js file at 935 (applyapptheme function) do i have to fix it or leave it as it is?

The more serious issue identified is DOM text reinterpreted as HTML without sanitizing (towards the end). Something should probably be done about that, maybe when values are first read.

THEBOSS0369 commented:
I will start my work on this pr now. Thanks for your patience 😃

THEBOSS0369 commented:
> OK, easiest is probably to hide the iframe contents until injection is done and stylesheet loaded. That requires the iframe itself to have a dark background colour by default, which is probably set by the background colour of the HTML page. That was the solution I adopted in the PWA, I seem to remember

Hello sir
I tried this method but doesn't seem to work i will try this issue for 2 - 3 day and will report whether it can be fixed or not.

THEBOSS0369 commented:
Hello @Jaifroid sir! i couldn't able to fix the main issue. The best i was able to do is the green color will show up then flash appear (for less than a sec) and then again green shows up and finally content loads properly. I don't think i will be able to fix it i tried every solution i can think of or search on ai & google but couldn't able to fix it 😔. 

For further let me know what to do with this pr whether it is acceptable or not. Also i have fixed the dom re represented as html issue , the codefactor error now is coming for complex method. Let me know what to do further with this pr .

THEBOSS0369 commented:
strange these ci tests should have passed

Jaifroid commented:
@THEBOSS0369 I'm underwater at the moment, but hope to get to this at the weekend! Rerunning tests...

Jaifroid commented:
@THEBOSS0369 As tests are systematically failing, I think it would be a good idea to roll back the latest commits (as you suggested elsewhere), and we should create a separate issue to deal with flash of incorrect style, so we can do final tests on this and merge your valuable work.

THEBOSS0369 commented:
Got it sir will do it before day after tomorrow.

THEBOSS0369 commented:
@Jaifroid sir the codefactor issue has been fixed and all test are passing except windows. If there is something else is required in this pr, let me know.

Jaifroid commented:
@THEBOSS0369 I notice that when I switch to the new dark mode, under some circumstances (can't quite tell when) I get a green background to the page, and this persists when I switch to light mode (see screenshot). If I refresh the app it disappears (in both modes), but comes back again after browsing to another article. Are you aware of having inserted a green background at any stage, maybe to mitigate against flash of incorrect style?

![image](https://github.com/user-attachments/assets/ca442f67-40db-4b77-9042-f5de2af4983a)

Jaifroid commented:
Excellent! I'll do a bit more testing and code review, and will let you know when it's ready to merge!

Jaifroid commented on www/index.html:
```diff
@@ -567,12 +567,14 @@ <h3 data-i18n="configure-display-settings-title">Display settings</h3>
                                         <option data-i18n="configure-display-themeoption-dark-apponly" value="dark">Dark (app only)</option>
                                         <option data-i18n="configure-display-themeoption-dark-invert" value="dark_invert">Dark (generic invert)</option>
                                         <option data-i18n="configure-display-themeoption-dark-mwinvert" value="dark_mwInvert">Dark (wikimedia invert)*</option>
+                                        <option data-i18n="configure-display-themeoption-dark-wikivector" id="theme-vector-option" title="Vector style only available for Wikimedia ZIMs" value="dark_wikiVector">Dark (Wikipedia Vector)*</option>
```
You need to add the translations for this in the i18n directory. You add the string `configure-display-themeoption-dark-wikivector` in the English file with "Dark (Wikipedia Vector)", and then you add the same strings for French and Spanish, but translating "Dark (Wikipedia Vector)" into those languages respectively. You can use AI for the translation and I'll check it as I know both languages.

THEBOSS0369 replied:
Done sir

Jaifroid commented on www/index.html:
```diff
@@ -567,12 +567,14 @@ <h3 data-i18n="configure-display-settings-title">Display settings</h3>
                                     </select>
                                     <p data-i18n="configure-display-apptheme-auto-description" class="text-muted" id="kiwix-auto-description" style="margin-bottom: 0;">&emsp;(Auto themes match the dark/light mode of your device.)</p>
                                     <p data-i18n="configure-display-apptheme-mwinvert-description" id="mwInvert-help" class="appThemeInfo form-text text-muted">* Implements workarounds specific to Wikimedia ZIMs. Try generic option if there are display errors with recent ZIMs.</p>
+                                    <p data-i18n="configure-display-apptheme-wikivector-description" id="wikiVector-help" class="appThemeInfo form-text text-muted">* Only applies to Wikimedia ZIMs. Falls back to generic inversion when other archives are loaded.</p>
```
Ditto

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -877,19 +888,28 @@ function applyAppTheme (theme) {
+    
+    // Safely handle help element IDs
+    var safeOldContentTheme = oldContentTheme.replace(/[^a-zA-Z0-9-]/g, '');
+    var safeContentTheme = contentTheme.replace(/[^a-zA-Z0-9-]/g, '');
```
Thanks for adding these safety features!

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -909,11 +929,54 @@ function applyAppTheme (theme) {
+                blockStyle.textContent = 'html, body { background-color: #e5e5e5 !important; }';
+                blockStyle.id = 'temp-dark-style';
+                doc.head.appendChild(blockStyle);
+            }
```
I think all the above code is part of the unsuccessful attempt to avoid Flash of Incorrect Style, right? In which case remove it, and we'll make a new issue to deal with that.

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -909,11 +929,54 @@ function applyAppTheme (theme) {
+                    }
+                }, 500);
+              }
+            };
```
Ditto, right? If not working the code should be removed.

Jaifroid reviewed (changes requested):
Just a few small changes from me, and any code that is part of the attempt to avoid Flash of Incorrect Style should be removed from here, as I'll probably revamp that part completely in a separate issue.

Jaifroid commented:
Also, please update branch from latest code on main using button below.

Jaifroid commented:
@THEBOSS0369 Just to let you know I intend to make a new app release this weekend. The changes needed here are all minor. Do you think you'd be able to complete them by Saturday? It would be nice to include this in the release. If not, no worries, it could go into the release after that.

THEBOSS0369 commented:
Got it sir i will try it do it tomorrow itself.

At first i thought that keeping this code might give u an idea of what i tried and what didn't worked. Though i will remove it now but if you want i can add all those unnecessary code in a google doc file, so that whenever you return to this issue you can know what have been tried already. 

Also good to see copilot making your work easy. This way it can help both maintainer and contributor to work really fast and effectively (although its just a start, but future seems interesting).... 🤯😈

Jaifroid reviewed (approved):
Thank you @THEBOSS0369 There are just some very minor tweaks to the translations -- you can just accept the suggestions for those if you like. Meanwhile I 'll do a test on another old browser and confirm before merge.

Jaifroid commented:
Thanks once again for this PR -- very useful, especially as the new ActionParse ZIMs coming out actually use Wikimedia Vector style, so it complements those well. 🙏🙏

Jaifroid commented:
Seems to have passed now!
